File: BubblerApp/BubblerApp/Features/Report/ReportPostViewModel.swift
import Foundation
import Combine

@MainActor
final class ReportPostViewModel: ObservableObject {
    @Published var selectedReason: ReportReason?
    @Published var comments = ""
    @Published var alsoBlockUser = false
    @Published var errorMessage: String?
    @Published private(set) var isSubmitting = false

    var canSubmit: Bool {
        selectedReason != nil && !isSubmitting
    }

    /// Cap and treat notes as untrusted plain text (backend enforces the same limit).
    func updateComments(_ value: String) {
        if value.count <= ReportDetailsLimits.maxLength {
            comments = value
        } else {
            comments = String(value.prefix(ReportDetailsLimits.maxLength))
        }
    }

    /// Report submit is UI-only until the review-queue API lands.
    /// When [alsoBlockUser] is on, the existing block endpoint is called immediately.
    func submit(blockUsername: String?, using authSession: AuthSession) async -> Bool {
        guard selectedReason != nil else {
            errorMessage = "Choose a reason before submitting."
            return false
        }

        if alsoBlockUser {
            let username = blockUsername?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            guard !username.isEmpty else {
                errorMessage = "We couldn't block this user."
                return false
            }

            isSubmitting = true
            errorMessage = nil
            defer { isSubmitting = false }

            do {
                _ = try await APIClient.blockUser(username: username)
            } catch {
                if case APIClientError.unauthorized = error {
                    authSession.signOut()
                }
                let description = error.localizedDescription.trimmingCharacters(in: .whitespacesAndNewlines)
                errorMessage = description.isEmpty
                    ? "We couldn't block this user."
                    : description
                return false
            }
        }

        errorMessage = nil
        return true
    }
}

File: BubblerAndroid/app/src/main/java/com/bubbler/android/features/report/ReportPostViewModel.kt
package com.bubbler.android.features.report

import com.bubbler.android.core.network.ApiException
import com.bubbler.android.data.model.ReportReason
import com.bubbler.android.data.repository.BlocksRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Report-post form state — mirrors Swift `ReportPostViewModel`.
 *
 * Report submit is UI-only until the review-queue API lands.
 * When [alsoBlockUser] is on, the existing block endpoint is called immediately.
 */
class ReportPostViewModel {
    private val _selectedReason = MutableStateFlow<ReportReason?>(null)
    val selectedReason: StateFlow<ReportReason?> = _selectedReason.asStateFlow()

    private val _comments = MutableStateFlow("")
    val comments: StateFlow<String> = _comments.asStateFlow()

    private val _alsoBlockUser = MutableStateFlow(false)
    val alsoBlockUser: StateFlow<Boolean> = _alsoBlockUser.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _isSubmitting = MutableStateFlow(false)
    val isSubmitting: StateFlow<Boolean> = _isSubmitting.asStateFlow()

    val canSubmit: Boolean
        get() = _selectedReason.value != null && !_isSubmitting.value

    fun selectReason(reason: ReportReason) {
        _selectedReason.value = reason
        _errorMessage.value = null
    }

    fun updateComments(value: String) {
        _comments.value = value
    }

    fun setAlsoBlockUser(value: Boolean) {
        _alsoBlockUser.value = value
    }

    /**
     * Report submit is UI-only. Blocking uses the live block API when requested.
     */
    suspend fun submit(
        authorUsername: String? = null,
        blocksRepository: BlocksRepository? = null,
        onUnauthorized: () -> Unit = {},
    ): Boolean {
        if (_selectedReason.value == null) {
            _errorMessage.value = "Choose a reason before submitting."
            return false
        }

        if (_alsoBlockUser.value) {
            val username = authorUsername?.trim().orEmpty()
            if (username.isEmpty() || blocksRepository == null) {
                _errorMessage.value = "We couldn't block this user."
                return false
            }

            _isSubmitting.value = true
            _errorMessage.value = null
            try {
                blocksRepository.blockUser(username)
            } catch (_: ApiException.Unauthorized) {
                onUnauthorized()
                _errorMessage.value = "We couldn't block this user."
                return false
            } catch (error: Exception) {
                val description = error.message?.trim().orEmpty()
                _errorMessage.value = description.ifEmpty { "We couldn't block this user." }
                return false
            } finally {
                _isSubmitting.value = false
            }
        }

        _errorMessage.value = null
        return true
    }
}
